cmd/kube-proxy/app/server_others.go

```diff
@@ -274,18 +274,14 @@ func getProxyMode(proxyMode string, iptver iptables.IPTablesVersioner, ipsetver
 		return proxyModeUserspace
 	}
 
-	if len(proxyMode) > 0 && proxyMode == proxyModeIPTables {
+	if proxyMode == proxyModeIPTables {
 		return tryIPTablesProxy(iptver, kcompat)
 	}
 
 	if utilfeature.DefaultFeatureGate.Enabled(features.SupportIPVSProxyMode) {
-		if proxyMode == proxyModeIPVS {
-			return tryIPVSProxy(iptver, ipsetver, kcompat)
-		} else {
-			glog.Warningf("Can't use ipvs proxier, trying iptables proxier")
-			return tryIPTablesProxy(iptver, kcompat)
-		}
+		return tryIPVSProxy(iptver, ipsetver, kcompat)
```

Comment:
It's not correct.  Enable feature gateway IPVS != specify IPVS proxy mode.

Reply:
Thanks for pointing it out! Since you have all of the IPVS context, could you please explain a bit more about what exactly does `enable feature gateway IPVS` mean?

Reply:
As I mentioned below, people can enable IPVS feature gateway, but it doesn't mean they always want to use IPVS proxy mode. Only if `feature-gateway-ipvs=true && proxy-mode=ipvs`, will IPVS-based kube-proxy start.

Reply:
From a user's perspective, I'm still confused by this. If I don't want to use IPVS proxy mode, why should I enable IPVS feature gate? In other words, if I enable the feature gate, in what scenario do I want to use IPVS mode and in what scenario don't I want to use IPVS mode? Thanks!

Reply:
If this PR change doesn't make sense, I guess we should at least document this a bit clear. (I didn't find any user guide docs clarifying this. Maybe I missed some.)

Reply:
IPVS-based kube-proxy is still in alpha phase, so we build a feature gateway in case of people use it by accident. Specifying `proxy-mode=ipvs` is not enough, people should enable IPVS alpha feature gateway by manual to make sure that they know all the potential risks.

Reply:
> I didn't find any user guide docs clarifying this.

xref https://github.com/kubernetes/kubernetes/tree/master/pkg/proxy/ipvs. Of course, if it doesn't work for you, you can send a PR...

Reply:
IIUC, what you say is also means that users should specify both the gate and the ipvs mode if they DO want to use IPVS mode, right?

Reply:
Yeah, exactly.

Reply:
@m1093782566 I think I get your point now. `feature-gateway-ipvs=true` means kube-proxy has the ability to support ipvs mode but users aren't limited to use ipvs mode only, right?

However, OTOH, we still should check if the feature gate is on when users want to use ipvs mode, i.e., `(!strings.Contains(config.FeatureGates, ipvsFeatureGate) && config.Mode == kubeproxyconfig.ProxyModeIPVS)` in line 149. Does this make sense?

Reply:
> `(!strings.Contains(config.FeatureGates, ipvsFeatureGate) && config.Mode == kubeproxyconfig.ProxyModeIPVS)` in line 149

@xiangpengzhao  Did you mean [179 line](https://github.com/xiangpengzhao/kubernetes/blob/dbb021b915412fc2a04087548b65dc81f177b61b/cmd/kube-proxy/app/server_others.go#L179) :)

I think if ipvs is well gated in `getProxyMode()`, it may not be necessary to have duplicate checks outside (for deciding which proxy mode to use). In fact, the original logic for this specific if condition seems reasonable to me.

Reply:
@MrHohn 

> (!strings.Contains(config.FeatureGates, ipvsFeatureGate) && config.Mode == kubeproxyconfig.ProxyModeIPVS) in line 149

xref: https://github.com/kubernetes/kubernetes/pull/56319/commits/416751484afd38d3ba13d0eb3f93b5f4194a6349#diff-efc74f26d8f8af5bdfbedf621e01b41dR147

I think the validation is not correct. People can enable IPVS feature gateway, but it doesn't mean they always want to use IPVS proxy mode. Only if `feature-gateway-ipvs=true && proxy-mode=ipvs`, will IPVS-based kube-proxy start.

Reply:
What's more, people can enable ALL alpha feature gateway, does it mean people always want to start IPVS-based kube-proxy even though they don't know the exists of IPVS.